src/runtime/StateSerialization/MaybeMethodBase.cs

```diff
@@ -0,0 +1,199 @@
+            }
+        }
+
+        public T Value
+        {
+            get
+            {
+                if (info == null)
+                {
+                    throw new SerializationException(DeletedMessage, innerException: deserializationException);
+                }
+                return (T)info;
+            }
+        }
+
+        public T UnsafeValue { get { return (T)info; } }
+        public string Name {get{return name;}}
+        public bool Valid => info != null;
+
+        public override string ToString()
+        {
+            return (info != null ? info.ToString() : $"missing method info: {name}");
+        }
+
+        public MaybeMethodBase(T mi)
+        {
+            info = mi;
+            name = mi?.ToString();
+            deserializationException = null;
+        }
+
+        internal MaybeMethodBase(SerializationInfo serializationInfo, StreamingContext context)
+        {
+            name = serializationInfo.GetString(SerializationName);
+            info = null;
+            deserializationException = null;
+            try
+            {
+                // Retrieve the reflected type of the method;
+                var typeName = serializationInfo.GetString(SerializationType);
+                var tp = Type.GetType(typeName);
+                if (tp == null)
+                {
+                    throw new SerializationException($"The underlying type {typeName} can't be found");
+                }
+                // Get the method's parameters types
+                var field_name = serializationInfo.GetString(SerializationMethodName);
+                var param = (ParameterHelper[])serializationInfo.GetValue(SerializationParameters, typeof(ParameterHelper[]));
+                Type[] types = new Type[param.Length];
+                bool hasRefType = false;
+                for (int i = 0; i < param.Length; i++)
+                {
+                    var paramTypeName = param[i].Name;
+                    types[i] = Type.GetType(paramTypeName);
+                    if (types[i] == null)
+                    {
+                        throw new SerializationException($"The parameter of type {paramTypeName} can't be found");
+                    }
+                    else if (types[i].IsByRef)
+                    {
+                        hasRefType = true;
+                    }
+                }
+
+                MethodBase mb = null;
+                if (serializationInfo.GetBoolean(SerializationIsCtor))
+                {
+                    // We never want the static constructor.
+                    mb = tp.GetConstructor(ClassManager.BindingFlags&(~BindingFlags.Static), binder:null, types:types, modifiers:null);
+                }
+                else
+                {
+                    mb = tp.GetMethod(field_name, ClassManager.BindingFlags, binder:null, types:types, modifiers:null);
```

Comment:
Considering you only get one method, how would it distinguish between `M(out int x)` and `M(ref int x)` or `M(in int x)`?

Reply:
`GetMethod` and `GetConstructor` can't. That's why there's a `CheckRefTypes` method in this struct. The `out_to_ref_param` tests this kind of scenario.

Reply:
But in this code you either receive 1 constructor or method, or none at all (an exception?). So if both `ref` and `out` are present, which one will it be?

Reply:
It throws an `AmbigousMatchException` if it resolves to more than one method/constructor. It's not possible to overload only with `in`/`ref`/`out` parameters. However, if in domain `1` a method has a `out` parameter and in domain `2`, same method, same parameter is now a `ref` parameter, it'll still bind.